### PR TITLE
feat: chain abstraction kotlin bindings

### PR DESCRIPTION
**Pull request by jakubuid**
- Kotlin bindings for Chain Abstraction

### CODE REVIEW

File: crates/yttrium/src/chain_abstraction/api/mod.rs [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 pub mod status;
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "uniffi", derive(uniffi_macros::Record))]
 #[serde(rename_all = "camelCase")]
 pub struct Transaction {
-    pub from: Address,
-    pub to: Address,
+    pub from: String,
+    pub to: String,
```
Comment (chris13524) [inline crates/yttrium/src/chain_abstraction/api/mod.rs:11]: Let's switch this back before merging

File: crates/yttrium/src/chain_abstraction/api/route.rs [hunk 2]
```diff
@@ -18,22 +16,25 @@ pub struct RouteRequest {
...
+    pub check_in: u32,
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "uniffi", derive(uniffi_macros::Record))]
 #[serde(rename_all = "camelCase")]
 pub struct FundingMetadata {
     pub chain_id: String,
-    pub token_contract: Address,
+    pub token_contract: String,
```
Comment (chris13524) [inline crates/yttrium/src/chain_abstraction/api/route.rs:31]: Switch back before merging

File: crates/yttrium/src/chain_abstraction/api/route.rs [hunk 3]
```diff
@@ -42,13 +43,19 @@ pub struct RouteResponseAvailable {
...
+#[cfg_attr(feature = "uniffi", derive(uniffi_macros::Record))]
+pub struct Empty;
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "uniffi", derive(uniffi_macros::Record))]
 #[serde(rename_all = "camelCase")]
 pub struct RouteResponseNotRequired {
     #[serde(rename = "transactions")]
-    _flag: [Infallible; 0],
+    _flag: Empty,
```
Comment (chris13524) [inline crates/yttrium/src/chain_abstraction/api/route.rs:54]: Does this work? I would expect you need `[Empty; 0]`?

File: crates/yttrium/src/chain_abstraction/api/route.rs [hunk 4]
```diff
@@ -18,22 +16,25 @@ pub struct RouteRequest {
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "uniffi", derive(uniffi_macros::Record))]
 #[serde(rename_all = "camelCase")]
 pub struct Metadata {
     pub funding_from: Vec<FundingMetadata>,
-    pub check_in: usize,
+    pub check_in: u32,
```
Comment (chris13524) [inline crates/yttrium/src/chain_abstraction/api/route.rs:23]: I feel like we should use `u64` instead here. That's what [`Duration::from_millis()`](https://doc.rust-lang.org/std/time/struct.Duration.html#method.from_millis) accepts

File: crates/yttrium/src/chain_abstraction/api/status.rs [hunk 5]
```diff
@@ -11,27 +11,31 @@ pub struct StatusQueryParams {
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "uniffi", derive(uniffi_macros::Record))]
 #[serde(rename_all = "camelCase")]
 pub struct StatusResponseSuccessPending {
-    created_at: usize,
+    created_at: u32,
```
Comment (chris13524) [inline crates/yttrium/src/chain_abstraction/api/status.rs:17]: Same for all of these, `u64` instead IMO

File: crates/yttrium/src/chain_abstraction/api/route.rs (hunk 4 above)
Comment (chris13524) [reply crates/yttrium/src/chain_abstraction/api/route.rs:23]: Actually we could go to `Duration` soon as uniffi supports that. We just have to make sure serde does too

File: crates/yttrium/src/chain_abstraction/api/route.rs (hunk 3 above)
Comment (jakubuid) [reply crates/yttrium/src/chain_abstraction/api/route.rs:54]: It works but will try the above

File: crates/kotlin-ffi/Cargo.toml [hunk 6]
```diff
@@ -8,13 +8,15 @@ name = "uniffi_yttrium"
 crate-type = ["cdylib"]
 
 [build-dependencies]
-uniffi_build = "0.28.1"
+uniffi_build = "0.28.3"
```
Comment (chris13524) [inline crates/kotlin-ffi/Cargo.toml:11]: Can we use git ref here too?

File: crates/kotlin-ffi/Cargo.toml [hunk 7]
```diff
@@ -8,13 +8,15 @@ name = "uniffi_yttrium"
...
 
 [dependencies]
 yttrium = { path = "../yttrium", features = ["uniffi"] }
-uniffi = { version = "0.28.1", features = ["tokio", "cli"] }
+uniffi = { git = "https://github.com/mozilla/uniffi-rs", features = ["tokio", "cli"] }
```
Comment (chris13524) [inline crates/kotlin-ffi/Cargo.toml:15]: Can we pin all of the uniffi deps to a commit ref with `rev`? https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#choice-of-commit

**Pull request merged**